app/services/merge_exact_duplicate_patients.rb

```diff
@@ -0,0 +1,56 @@
+class MergeExactDuplicatePatients
+  include Memery
+
+  def initialize
+    @merge_failures = []
+  end
+
+  attr_accessor :merge_failures
+
+  def perform
+    duplicate_patient_ids.map do |patient_ids|
+      DeduplicatePatients.new(Patient.where(id: patient_ids)).merge
+    rescue => e
+      # Bad data can cause our merge logic to breakdown in unpredictable ways.
+      # We want to report any such errors and look into them on a per case basis.
+      handle_error(e, patient_ids)
+    end
+
+    report
+  end
+
+  def handle_error(e, patient_ids)
+    merge_failures << {exception: e, patient_ids: patient_ids}
+
+    Rails.logger.debug "Failed to merge patients #{patient_ids}"
+    Sentry.capture_message("Failed to merge duplicate patients", extra: {exception: e, patient_ids: patient_ids})
+  end
+
+  def report_stats
+    {processed: {total: duplicate_patient_ids.flatten.count,
+                 distinct: duplicate_patient_ids.count},
+     merged: {total: duplicate_patient_ids.flatten.count - merge_failures.flatten.count,
+              distinct: duplicate_patient_ids.count - merge_failures.count,
+              total_failures: merge_failures.count,
+              distinct_failure: merge_failures.flatten.count}}
+  end
+
+  def report
+    Rails.logger.info(report_stats.to_json)
+    Rails.logger.info "Failed to merge patients #{merge_failures}"
```

Comment:
The logging seems a bit convoluted to me. `merge_failures` is populated in the `rescue` block, reported in `handle_error`, but also reported here? Could we tighten this up with

```ruby
	    duplicate_patient_ids.map do |patient_ids|
      DeduplicatePatients.new(Patient.where(id: patient_ids)).merge
      report
    rescue => e
      # Bad data can cause our merge logic to breakdown in unpredictable ways.
      # We want to report any such errors and look into them on a per case basis.
      handle_error(e, patient_ids)
      report_failures
    end
```